hydra-node/src/Hydra/Network/Etcd.hs

```diff
@@ -0,0 +1,513 @@
+
+putLastKnownRevision :: MonadIO m => FilePath -> Natural -> m ()
+putLastKnownRevision directory rev = do
+  liftIO $ encodeFile (directory </> "last-known-revision.json") rev
+
+-- | Write a well-known key to indicate being alive, keep it alive using a lease
+-- and poll other peers enries to yield connectivity events. While doing so,
+-- overall network connectivity is determined from the ability to read/write to
+-- the cluster.
+pollConnectivity ::
+  Connection ->
+  -- | Local host
+  Host ->
+  NetworkCallback msg IO ->
+  IO ()
+pollConnectivity conn localHost NetworkCallback{onConnectivity} = do
+  seenAliveVar <- newTVarIO []
+  withGrpcContext "pollConnectivity" . forever $ do
+    leaseId <- createLease
+    -- If we can create a lease, we are connected
+    onConnectivity NetworkConnected
+    handle onGrpcException $ do
+      -- Write our alive key using lease
+      writeAlive leaseId
+      withKeepAlive leaseId $ \keepAlive ->
+        forever $ do
+          -- Keep our lease alive
+          keepAlive
+          -- Determine alive peers
+          alive <- getAlive
+          let othersAlive = alive \\ [localHost]
+          seenAlive <- atomically $ swapTVar seenAliveVar othersAlive
+          forM_ (othersAlive \\ seenAlive) $ onConnectivity . PeerConnected
+          forM_ (seenAlive \\ othersAlive) $ onConnectivity . PeerDisconnected
+          threadDelay 1
+ where
+  ttl = 3
+
+  onGrpcException GrpcException{grpcError = GrpcUnavailable} = do
+    onConnectivity NetworkDisconnected
+    threadDelay 1
+  onGrpcException e = throwIO e
+
+  -- REVIEW: server can decide ttl?
+  createLease = withGrpcContext "createLease" $ do
+    leaseResponse <-
+      nonStreaming conn (rpc @(Protobuf Lease "leaseGrant")) $
+        defMessage & #ttl .~ ttl
+    pure $ leaseResponse ^. #id
+
+  writeAlive leaseId = withGrpcContext "writeAlive" $ do
+    void . nonStreaming conn (rpc @(Protobuf KV "put")) $
+      defMessage
+        & #key .~ "alive-" <> show localHost
+        & #value .~ serialize' localHost
+        & #lease .~ leaseId
+
+  withKeepAlive leaseId action = do
+    biDiStreaming conn (rpc @(Protobuf Lease "leaseKeepAlive")) $ \send _recv ->
+      void . action $ send $ NextElem (defMessage & #id .~ leaseId)
+
+  getAlive = withGrpcContext "getAlive" $ do
+    res <-
+      nonStreaming conn (rpc @(Protobuf KV "range")) $
+        defMessage
+          & #key .~ "alive"
+          & #rangeEnd .~ "alivf" -- NOTE: e+1 to query prefixes
+    pure $ flip mapMaybe (res ^.. #kvs . traverse . #value) $ \bs ->
+      -- XXX: Silently swallow incompatible values. Hard to debug, but failure
+      -- to decode here should not crash the component either.
+      case decodeFull' bs of
+        Left _err -> Nothing
+        Right x -> pure x
+
+-- | Add context to the 'grpcErrorMessage' of any 'GrpcException' raised.
+withGrpcContext :: MonadCatch m => Text -> m a -> m a
```

Comment:
I'd prefer to not edit thrown exceptions from a third party library, rather construct new ones with more context if needed.

Reply:
I prefer less code, but will also not fight over it. Feel free to do the change in a follow-up.